velox/exec/prefixsort/PrefixSortAlgorithm.h

```diff
@@ -0,0 +1,343 @@
+    prefix_ += entrySize_;
+    return *this;
+  }
+
+  inline PrefixSortIterator& operator--() {
+    prefix_ -= entrySize_;
+    return *this;
+  }
+
+  inline PrefixSortIterator operator++(int) {
+    auto tmp = *this;
+    prefix_ += entrySize_;
+    return tmp;
+  }
+
+  inline PrefixSortIterator operator--(int) {
+    auto tmp = *this;
+    prefix_ -= entrySize_;
+    return tmp;
+  }
+
+  inline PrefixSortIterator operator+(const uint64_t i) const {
+    auto result = *this;
+    result.prefix_ += i * entrySize_;
+    return result;
+  }
+
+  inline PrefixSortIterator operator-(const uint64_t i) const {
+    PrefixSortIterator result = *this;
+    result.prefix_ -= i * entrySize_;
+    return result;
+  }
+
+  inline PrefixSortIterator& operator=(const PrefixSortIterator& other) {
+    VELOX_CHECK_EQ(entrySize_, other.entrySize_);
+    prefix_ = other.prefix_;
+    return *this;
+  }
+
+  inline uint64_t operator-(const PrefixSortIterator& other) const {
+    VELOX_CHECK_EQ((prefix_ - other.prefix_) % entrySize_, 0);
+    VELOX_CHECK_GE(prefix_ - other.prefix_, 0);
+    return (prefix_ - other.prefix_) / other.entrySize_;
+  }
+
+  inline bool operator<(const PrefixSortIterator& other) const {
+    return prefix_ < other.prefix_;
+    ;
+  }
+
+  inline bool operator>(const PrefixSortIterator& other) const {
+    return prefix_ > other.prefix_;
+    ;
+  }
+
+  inline bool operator>=(const PrefixSortIterator& other) const {
+    return prefix_ >= other.prefix_;
+    ;
+  }
+
+  inline bool operator<=(const PrefixSortIterator& other) const {
+    return prefix_ <= other.prefix_;
+    ;
+  }
+
+  inline bool operator==(const PrefixSortIterator& other) const {
+    return prefix_ == other.prefix_;
+    ;
+  }
+
+  inline bool operator!=(const PrefixSortIterator& other) const {
+    return prefix_ != other.prefix_;
+  }
+
+ private:
+  char* prefix_;
+  const uint64_t entrySize_;
+};
+
+/// Provides a series of methods (mostly required by the sort
+/// algorithm) for PrefixSort. Maintains the buffer for swap operations and
+/// some necessary context information such as entrySize.
+class PrefixSortContext {
+ public:
+  PrefixSortContext(memory::MemoryPool* pool, uint64_t entrySize, char* end);
+
+  inline void swap(const PrefixSortIterator& lhs, const PrefixSortIterator& rhs)
+      const {
+    VELOX_CHECK_LT(*lhs, end_);
+    VELOX_CHECK_LT(*rhs, end_);
+    memcpy(rawSwapBuffer_, *lhs, entrySize_);
+    memcpy(*lhs, *rhs, entrySize_);
+    memcpy(*rhs, rawSwapBuffer_, entrySize_);
+  }
+
+  inline void rangeSwap(
+      const PrefixSortIterator& start1,
+      const PrefixSortIterator& start2,
+      uint64_t length) const {
+    VELOX_CHECK_LE(*(start1 + length), end_);
+    VELOX_CHECK_LE(*(start2 + length), end_);
+    char* d1 = *start1;
+    char* d2 = *start2;
+    for (uint64_t i = 0; i < length; i++, d1 += entrySize_, d2 += entrySize_) {
+      // 1. Use swap() method above like swap(start1 +i , start2 + i) will slow
```

Comment:
>  will slow down performance.

Is this accurate? I thought this would be the case only if swap doesn't end up inlined. Let's clarify.